cmd/kube-proxy/app/server.go

```diff
@@ -604,21 +607,37 @@ func (s *ProxyServer) Run() error {
 	}
 
 	informerFactory := informers.NewSharedInformerFactoryWithOptions(s.Client, s.ConfigSyncPeriod,
-		informers.WithTweakListOptions(func(options *v1meta.ListOptions) {
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
 			options.LabelSelector = "!" + apis.LabelServiceProxyName
 		}))
 
-	// Create configs (i.e. Watches for Services and Endpoints)
+	// Create configs (i.e. Watches for Services and Endpoints or EndpointSlices)
 	// Note: RegisterHandler() calls need to happen before creation of Sources because sources
 	// only notify on changes, and the initial update (on process start) may be lost if no handlers
 	// are registered yet.
 	serviceConfig := config.NewServiceConfig(informerFactory.Core().V1().Services(), s.ConfigSyncPeriod)
 	serviceConfig.RegisterEventHandler(s.Proxier)
 	go serviceConfig.Run(wait.NeverStop)
 
-	endpointsConfig := config.NewEndpointsConfig(informerFactory.Core().V1().Endpoints(), s.ConfigSyncPeriod)
-	endpointsConfig.RegisterEventHandler(s.Proxier)
-	go endpointsConfig.Run(wait.NeverStop)
+	if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSlice) {
+		endpointSliceConfig := config.NewEndpointSliceConfig(informerFactory.Discovery().V1alpha1().EndpointSlices(), s.ConfigSyncPeriod)
+		endpointSliceConfig.RegisterEventHandler(s.Proxier)
+		go endpointSliceConfig.Run(wait.NeverStop)
+
+		sliceList, err := s.Client.DiscoveryV1alpha1().EndpointSlices("").List(metav1.ListOptions{})
```

Comment:
The goal here was to initialize the local cache with all endpoint slices. If I'm remembering correctly, "" would fetch resources for all namespaces. As far as checking to see if the API was enabled, I had been assuming if someone had set the feature gate here they would have also enabled the API, but it makes sense to add some error handling here if that's not the case.

Reply:
You do not need to list it on start up. It will relist and come in as update event upon restart. 

You can handle Add/Update/Delete separately to populate the cache.

Reply:
Thanks, lesson learned, had thought that we'd only get new events after the informer had initialized.